src/include/mempool.h

```diff
@@ -27,6 +27,7 @@
 #include <boost/container/flat_map.hpp>
 
 #if defined(_GNU_SOURCE) && defined(WITH_SEASTAR) && !defined(WITH_ALIEN)
+#  define MEMPOOL_SCHED_GETCPU
```

Comment:
Is there a chance to simplify? In my mind the main worry about `sched_getcpu()` is a platform-we-do-support where it (or similar mechanism) is simply unavailable. I'm aware there are some FreeBSD users running Ceph.  Still, this was based more on pure concern than actual research.

What do you think about making `sched_getcpu()` the main path also for the classical world while providing a (nicely encapsulated) fallback based on the thread2shard mapping?

Reply:
The 2nd commit just refactors existing code so keeps the same behavior as the existing code of only use sched_getcpu() if _GNU_SOURCE, WITH_SEASTART and !WITH_ALIEN are defined, falling back to assigning threads to shards in a round robin fashion otherwise. Round robin sharding is more portable than using a hash of pthread_self() and fixes an issue seen with this hashing on aarch64 platforms.

The 3rd commit builds on this and uses sched_getcpu() for all builds where _GNU_SOURCE is defined, with adaptive caching to deal with the fact that for classic builds threads are not bound to cores. This means classic (and Crimson + Bluestore) builds get most of the performance benefit that Crimson builds were getting. FreeBSD (and other platforms) without sched_getcpu() fall back to the thread sharding.

Scaling the number of shards according to the number of CPU cores means that even with thread sharding the risk of sharing caching lines will be low on high end servers with a lot of cores.